### PR TITLE
Refactor and test `BuildContext`

### PR DESCRIPTION
**Pull request by paulbalaji**
#### Description

Slight refactor of `BuildContext`.

- Introduce `BuildContextFilter` struct that is passed to `GetBuildContexts`.
- Slightly refactored build menu methods.
- Also includes a fix for https://improbableio.atlassian.net/browse/UTY-2689.

#### Tests

- yes there are tests

#### Documentation

- [x] changelog

#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.

### CODE REVIEW

File: workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildContext.cs [hunk 1]
```diff
@@ -14,60 +12,30 @@ public struct BuildContext
         public ScriptingImplementation ScriptingImplementation;
         public iOSSdkVersion? IOSSdkVersion;
 
-        public static List<BuildContext> GetBuildContexts(IEnumerable<string> wantedWorkerTypes,
-            BuildEnvironment buildEnvironment, ScriptingImplementation? scriptImplementation = null,
-            ICollection<BuildTarget> buildTargetFilter = null, iOSSdkVersion? iosSdkVersion = null)
+        public static List<BuildContext> GetBuildContexts(BuildContextFilter contextFilter)
         {
             var spatialOsBuildConfiguration = BuildConfig.GetInstance();
```
Comment (jamiebrynes7) [inline workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildContext.cs:17]: In order to test properly, this should be injected as well right? Currently the tests will use whatever build configuration is defined in the project which: 

- doesn't allow us to construct edge cases to test
- means our tests are brittle and can break because we changed the default build configuration

File: workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildEnvironmentConfig.cs [hunk 2]
```diff
@@ -43,5 +45,45 @@ public BuildEnvironmentConfig(IEnumerable<BuildTarget> availableTargets, params
...
+                .ToList();
+
+            // Error on missing required build support
+            if (missingTargets.Any(c => c.Required))
+            {
+                var targetNames = string.Join(", ", missingTargets
+                    .Where(c => c.Required)
+                    .Select(c => c.Target.ToString()));
+                throw new BuildFailedException(
+                    $"Build failed for {workerType}. Cannot build for required ({targetNames}) because build support is not installed in the Unity Editor.");
```
Comment (jamiebrynes7) [inline workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildEnvironmentConfig.cs:73]: Seems a bit like a code smell that we pass in the `workerType` just to add context to this exception. 

The caller could equally add this context right?

File: workers/unity/Packages/io.improbable.gdk.buildsystem/Tests/BuildContextTests.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        [TestCase("UnityGameLogic", 1, BuildTarget.StandaloneWindows64)]
+        [TestCase("DummyWorkerType", 0, BuildTarget.StandaloneWindows64)]
+#elif UNITY_EDITOR_OSX
+        [TestCase("UnityClient", 1, BuildTarget.StandaloneOSX)]
+        [TestCase("UnityGameLogic", 1, BuildTarget.StandaloneOSX)]
+        [TestCase("DummyWorkerType", 0, BuildTarget.StandaloneOSX)]
+#endif
+        [TestCase("UnityClient", 0, BuildTarget.StandaloneLinux64)]
+        [TestCase("UnityGameLogic", 0, BuildTarget.StandaloneLinux64)]
+        [TestCase("DummyWorkerType", 0, BuildTarget.StandaloneLinux64)]
```
Comment (jamiebrynes7) [inline workers/unity/Packages/io.improbable.gdk.buildsystem/Tests/BuildContextTests.cs:51]: If we inject the `BuildConfig` rather than using the project one we can avoid this

File: workers/unity/Packages/io.improbable.gdk.buildsystem/Tests/BuildContextTests.cs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        }
+
+#if UNITY_EDITOR_WIN
+        [TestCase("UnityClient", BuildTarget.StandaloneWindows64)]
+        [TestCase("UnityGameLogic", BuildTarget.StandaloneWindows64)]
+#elif UNITY_EDITOR_OSX
+        [TestCase("UnityClient", BuildTarget.StandaloneOSX)]
+        [TestCase("UnityGameLogic", BuildTarget.StandaloneOSX)]
+#endif
+        public void GetBuildContexts_for_valid_worker_type_without_filter_returns_all_expected_local_contexts(string workerType, params BuildTarget[] expectedTargets)
```
Comment (jamiebrynes7) [inline workers/unity/Packages/io.improbable.gdk.buildsystem/Tests/BuildContextTests.cs:94]: The `expectedTargets` parameter isn't used in the body of the test

File: workers/unity/Packages/io.improbable.gdk.buildsystem/WorkerBuilder.cs [hunk 5]
```diff
@@ -30,15 +30,17 @@ public static void Build()
...
-                var buildEnvironment = CommandlineParser.GetBuildEnvironment(args);
-                var targetIOSSdkVersion = CommandlineParser.GetTargetIOSSdk(args);
+                var buildContextFilter = new BuildContextFilter
+                {
+                    WantedWorkerTypes = CommandlineParser.GetWorkerTypesToBuild(args),
+                    BuildEnvironment = CommandlineParser.GetBuildEnvironment(args),
+                    ScriptImplementation = CommandlineParser.GetScriptingImplementation(args),
+                    BuildTargetFilter = CommandlineParser.GetBuildTargetFilter(args),
+                    IOSSdkVersion = CommandlineParser.GetTargetIOSSdk(args)
+                };
```
Comment (jamiebrynes7) [inline workers/unity/Packages/io.improbable.gdk.buildsystem/WorkerBuilder.cs:40]: Move this into a `BuildContextFilter.FromCommandLine(args);`? 

Also `CommandlineParser.GetWorkerTypesToBuild` and others can be moved in there too if they aren't used elsewhere (I don't believe they are).

**Comment by jamiebrynes7** [comment]
Also it would be nice to be able to test the build support checking part of the `GetBuildContexts()`. This would require more DI/mocking though

File: workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildContext.cs (hunk 1 above)
Comment (paulbalaji) [reply workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildContext.cs:17]: Correct - but I'm not sure how I'd inject our own BuildConfig if it's a singleton scriptable object

File: workers/unity/Packages/io.improbable.gdk.buildsystem/Tests/BuildContextTests.cs (hunk 4 above)
Comment (paulbalaji) [reply workers/unity/Packages/io.improbable.gdk.buildsystem/Tests/BuildContextTests.cs:94]: Ah yes I forgot something during a copypasta

File: workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildContext.cs (hunk 1 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildContext.cs:17]: The `SingletonScriptableObject` allows you to make new versions of the assets and doesn't count it as long as its not saved to disk (i.e. - in memory only).

https://github.com/spatialos/gdk-for-unity/blob/6b8e88650b7132c83120ce642e2fd39b7f63e846/workers/unity/Packages/io.improbable.gdk.core/Editor/SingletonScriptableObject.cs#L28

You should be able to use the `CreateInstance` method as normal if I'm looking at this correctly.

Comment (paulbalaji) [reply workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildContext.cs:17]: ah nice, I'll look into it

File: workers/unity/Packages/io.improbable.gdk.buildsystem/Tests/BuildContextTests.cs (hunk 3 above)
Comment (paulbalaji) [reply workers/unity/Packages/io.improbable.gdk.buildsystem/Tests/BuildContextTests.cs:51]: not really, this is testing the filters that we can pass through cli

Comment (jamiebrynes7) [reply workers/unity/Packages/io.improbable.gdk.buildsystem/Tests/BuildContextTests.cs:51]: I mean we don't need the ifdefs (since we can control the underlying `BuildConfig` object so its not OS dependent).

File: workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildContext.cs [hunk 6]
```diff
@@ -14,60 +13,37 @@ public struct BuildContext
...
                 result.AddRange(supportedTargets.Select(targetConfig => new BuildContext
                 {
                     WorkerType = workerType,
-                    BuildEnvironment = buildEnvironment,
-                    ScriptingImplementation = scriptImplementation ?? targetConfig.ScriptingImplementation,
+                    BuildEnvironment = contextFilter.BuildEnvironment,
+                    ScriptingImplementation = contextFilter.ScriptImplementation ?? targetConfig.ScriptingImplementation,
                     BuildTargetConfig = targetConfig,
-                    IOSSdkVersion = (targetConfig.Target == BuildTarget.iOS) ? iosSdkVersion : null
+                    IOSSdkVersion = (targetConfig.Target == BuildTarget.iOS) ? contextFilter.IOSSdkVersion : null
```
Comment (jamiebrynes7) [inline workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildContext.cs:46]: These make me feel `BuildContextFilter` is misnamed as this isn't filtering build targets, but actually applying certain settings to them.

File: workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildContextFilter.cs [hunk 7]
```diff
@@ -1,22 +1,91 @@
...
-        internal static string[] GetWorkerTypesToBuild(CommandLineArgs args)
+        public IEnumerable<string> WantedWorkerTypes;
+        public BuildEnvironment BuildEnvironment;
+        public ScriptingImplementation? ScriptImplementation;
+        public ICollection<BuildTarget> BuildTargetFilter;
+        public iOSSdkVersion? IOSSdkVersion;
+
+        public BuildContextFilter(IEnumerable<string> wantedWorkerTypes, BuildEnvironment buildEnvironment,
+            ScriptingImplementation? scriptImplementation = null, ICollection<BuildTarget> buildTargetFilter = null,
+            iOSSdkVersion? iosSdkVersion = null)
```
Comment (jamiebrynes7) [inline workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildContextFilter.cs:19]: We can probably drop these optional parameters from the constructor as you can set them with the initializer syntax if you wish:

```csharp
new BuildContextFilter(wantedWorkerTypes, buildEnvironment)
{
	ScriptImplementation = ScriptingImplementation.IL2CPP
}
```

File: workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildEnvironmentConfig.cs [hunk 8]
```diff
@@ -43,5 +45,44 @@ public BuildEnvironmentConfig(IEnumerable<BuildTarget> availableTargets, params
                 }
             }
         }
+
+        public IEnumerable<BuildTargetConfig> GetSupportedTargets(BuildContextFilter contextFilter)
+        {
+            // Filter targets for CI
+            var targetConfigs = BuildTargets
+                .Where(t => t.Enabled && (contextFilter.BuildTargetFilter?.Contains(t.Target) ?? true))
```
Comment (jamiebrynes7) [inline workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildEnvironmentConfig.cs:53]: Could add:

```
public bool Matches(BuildTarget target);
```

to the `BuildContextFilter` to slim this down

File: workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildConfig.cs [hunk 9]
```diff
@@ -7,15 +7,15 @@
 namespace Improbable.Gdk.BuildSystem.Configuration
 {
     [CreateAssetMenu(fileName = "SpatialOS Build Configuration", menuName = BuildConfigEditor.BuildConfigurationMenu)]
-    internal class BuildConfig : SingletonScriptableObject<BuildConfig>
+    public class BuildConfig : SingletonScriptableObject<BuildConfig>
```
[the remark this answers is not in the document]
Comment (paulbalaji) [reply workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildConfig.cs:10]: yes because you need to provide a `BuildConfig` to `GetBuildContexts`, which is public

File: workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildContext.cs (hunk 6 above)
Comment (paulbalaji) [reply workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildContext.cs:46]: yeah the naming came from the `BuildTargetFilter` field in this

`BuildContextSettings`?

Comment (jamiebrynes7) [reply workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildContext.cs:46]: Yeah sgtm

File: workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildContextFilter.cs (hunk 7 above)
Comment (paulbalaji) [reply workers/unity/Packages/io.improbable.gdk.buildsystem/Configuration/BuildContextFilter.cs]: I had to add custom setters for scripting/target properties, to make sure ios is always il2cpp
